### PR TITLE
Ensure that all exceptions are captured and returned to the callee

### PR DESCRIPTION
**Pull request by adamretter**
Previously when either nesting managed resources or using `and` to compose them, if an exception occurred in the close method of the outer managed resource, then any exceptions produced by the inner managed resource were lost.

1. The tests that I have added clearly show the issue

2. I have produced a fix; There are maybe other ways to fix this but the approach I settled on was pragmatic and minimally intrusive as far as I could figure out. Happy to discuss...

### CODE REVIEW

File: src/main/scala/resource/ExtractableManagedResource.scala [hunk 1]
```diff
@@ -46,5 +46,7 @@ trait ExtractableManagedResource[+A] extends ManagedResource[A] {
    *        An either where the left hand side is the currently contained resource unless exceptions, in which case
    *        the right hand side will contain the sequence of throwable encountered.
    */
-  def either: Either[Seq[Throwable], A]
+  def either: ExtractedEither[Seq[Throwable], A]
 }
+
+case class ExtractedEither[+A, +B](either: Either[A, B])
```
Comment (jsuereth) [inline src/main/scala/resource/ExtractableManagedResource.scala]: Is there a way to fix the bug without introducing this container?  I'm hesitant to significantly alter the return type of this guy without moving to `scala.util.Try` and using Java 8's "suppressed exception" mechanisms...

**Comment by jsuereth** [comment]
Tests look great and the fix looks good.  I'd love to merge as-is, but I'd like to try to not expose `ExtractedEither` in the public API, as well as possibly using scala.util.Try (new since ARM was first written) over Either...

File: src/main/scala/resource/ExtractableManagedResource.scala (hunk 1 above)
Comment (adamretter) [inline src/main/scala/resource/ExtractableManagedResource.scala]: As far as I can see there is no way to fix the bug without introducing either a container of a sub-type of Either. Introducing a container is significantly less effort than a subclass. and placing an implicit conversion in the package object (https://github.com/adamretter/scala-arm/commit/c7e31438b551f34536f1962ebcb1c5bfeb4b6148#diff-c1a4ce80b3e56af8d61bb72ceb8a46b7R75) means that no one using the API will be distrurbed as the `ExtractedEither` will always be unboxed implicitly for them to an `Either`.

To be honest having done a lot of work functional work with Java 8 and `try-with-resources` (which was actually introduced in Java 7), the *suppressed exception* mechanism is just terrible and can lead to very hard to debug problems. Especially if you consider that `try-with-resources` and `managed` are not just for opening and closing IO Streams, but for any object which implements `close`. I would be very disappointed if scala-arm made the same mistakes as Java's `try-with-resources`, you already have a much better mechanism.

**Comment by adamretter** [comment]
@jsuereth Understood. See my comment above... I don't think there is another simple way to do this, and the `ExtractedEither` is effectively invisible due to the implicit conversion.

If I used a subclass I could have the `.either` API indicate that it just returns an `Either` but in reality I would be returning a subclass of Either which wraps the original Either. I think subclassing Either for the purposes of wrapping will be quite a lot of work (could maybe be simplified with runtime proxy generation, but that would introduce dependencies on cglib etc.).

**Comment by adamretter** [comment]
@jsuereth Just to add that I think switching to `scala.util.Try` would be a mistake, as `Failure` only allows you a single exception, which is a very similar problem to what I am experiencing at the moment and lead to this PR.

**Comment by adamretter** [comment]
@jsuereth Any further thoughts or are you 100% committed to going the `scala.util.Try` and suppressed exceptions route?

**Comment by adamretter** [comment]
@jsuereth Any update please?

**Comment by jsuereth** [comment]
@adamretter Sorry for the LONG delay, I've been basically living in a hole (actually rewriting the scala pickles library has been consuming all my OSS time).

Ok, so my thoughts:

1. Your code is excellent, just FYI
2. I would still like to use Try
3. Java added 'suppressedExceptions' specifically to handle multiple exceptions.  We have an open PR to make use of that.
4. We *could* possibly use your ExtractedEither and provide a "Try" API on top of it.  I'm leaning towards that right now.

**Comment by adamretter** [comment]
@jsuereth Okay no problem. When you say:

>Java added 'suppressedExceptions' specifically to handle multiple exceptions.

Do you mean that with `try-with-resources` in Java, then some exceptions can be suppressed? or maybe there is some Java feature I have missed? If so, then I have to say that it was a terrible architecture choice and leads to very hard to debug issues in real world applications in my experience. I really must sort a blog post out on why suppressed expressions are the new `null`.

**Comment by jsuereth** [comment]
@adamretter That is what i was referring too.   If it's truly terrible I think we only offer a compatibility layer for people who really want it.  I'm ok using what we have.

Note: that Try already has a mechanism to aggregate exceptions used by parallel collections (which have a similar issue).   I'd like to use standard-library features where appropriate, because I'd like this library to be mostly standard and less 'resistance', but I'm ok making better decisions.

My big hesitation with Either is just that it's a terrible data structure as implemented in the standard library.   Either SHOULD be great, but it isn't that great in its currnet state.  It's very clunky and inelegant.

File: src/main/scala/resource/ExtractableManagedResource.scala (hunk 1 above)
Comment (jsuereth) [inline src/main/scala/resource/ExtractableManagedResource.scala]: Right.  SO I think what I'd like to see (possibly not this CL, possibly before release) is that IF we're not going to use Either itself, then we create our own wrapper of Either that is its own datastrcture.

Also, you do not need to create an implicit conversion in the package object, simply do so in the companion object ( a better place).


Can you move `ExtractedEither` into its own file, and possibly rename it to `ArmResult`, so we know it's not exactly an either.  I'm ok with "easing" transition to Either, but FYI sometimes these "easy" mechanisms actually aren't as nice as you would think.  E.g. pattern matching MAY not work the way we want it (if the Left/Right pattern match is against `Any` and not `Either`).   Lots of subtle bugs can occur here.

So TL;DR;

* I'm ok NOT using supressed exceptions
* We should rename ExtractedEither into something more useful.
   - We can add explicit `.toTry`, `.toEither` methods to it
   - It's ok to break compatibility in the 2.0 release.  I'd rather do this than implicit conversion from `ExtractedEither => Either`, as this is too dangerous when we expect pattern matching.


Now the question is, since you've done all the work, would you like to flesh out a "ArmResult" (a.k.a. ExtractedEither) any further?

**Comment by jsuereth** [comment]
Ok, merging this for now, we can improve `ExtractedEither` in follow on commits.

**Comment by adamretter** [comment]
@jsuereth Hooray, thank you :-) :-)
